source/h2_connection.c

```diff
@@ -368,6 +381,64 @@ static void s_handler_destroy(struct aws_channel_handler *handler) {
     aws_mem_release(connection->base.alloc, connection);
 }
 
+static struct h2_pending_settings *s_new_pending_settings(
+    struct aws_allocator *allocator,
+    const struct aws_h2_frame_setting *settings_array,
+    size_t num_settings) {
+
+    size_t settings_storage_size = sizeof(struct aws_h2_frame_setting) * num_settings;
+    struct h2_pending_settings *pending_settings;
+    void *settings_storage;
+    if (!aws_mem_acquire_many(
+            allocator,
+            2,
+            &pending_settings,
+            sizeof(struct h2_pending_settings),
+            &settings_storage,
+            settings_storage_size)) {
+        return NULL;
+    }
+
+    AWS_ZERO_STRUCT(*pending_settings);
+    /* We buffer the settings up, incase the caller has freed them when the ACK arrives */
+    pending_settings->settings_array = settings_storage;
+    memcpy(pending_settings->settings_array, settings_array, num_settings * sizeof(*settings_array));
+    pending_settings->num_settings = num_settings;
+
+    return pending_settings;
+}
+
+int aws_h2_connection_change_settings(
+    struct aws_h2_connection *connection,
+    const struct aws_h2_frame_setting *settings_array,
+    size_t num_settings) {
+    AWS_PRECONDITION(aws_channel_thread_is_callers_thread(connection->base.channel_slot->channel));
+    AWS_PRECONDITION(settings_array);
+
+    if (!num_settings) {
+        return AWS_OP_SUCCESS;
+    }
+
+    /* push the setting array into the queue, not applying the change until it is ACKed by peer */
+    struct h2_pending_settings *pending_settings =
+        s_new_pending_settings(connection->base.alloc, settings_array, num_settings);
+    if (!pending_settings) {
+        return AWS_OP_ERR;
+    }
+    /* Send setting frame to inform our peer */
+    struct aws_h2_frame *setting_frame =
+        aws_h2_frame_new_settings(connection->base.alloc, settings_array, num_settings, false /*ACK*/);
+    if (!setting_frame) {
+        CONNECTION_LOGF(ERROR, connection, "Failed to send setting_frames, error %s", aws_error_name(aws_last_error()));
+        aws_mem_release(connection->base.alloc, pending_settings);
+        return AWS_OP_ERR;
+    }
+    aws_h2_connection_enqueue_outgoing_frame(connection, setting_frame);
+
+    aws_linked_list_push_back(&connection->thread_data.pending_settings_queue, &pending_settings->node);
+    return AWS_OP_SUCCESS;
```

Comment:
If aws_h2_connection_change_settings() is going to be public-ish, let's have it s_try_write_outgoing_frames() at the end